htmresearch/algorithms/image_sparse_net.py

```diff
@@ -0,0 +1,182 @@
+
+    return images
+
+
+  def loadNumpyImages(self, path, key=None):
+    """
+    Loads images using numpy.
+
+    :param path:      (string)   Path to data file
+    :param key:       (string)   Object key in data file if it's a dict
+
+    Also stores image dimensions to later the original images. If there are
+    multiple channels, self.numChannels will store the number of channels,
+    otherwise it will be set to None.
+    """
+    data = np.load(path)
+
+    if isinstance(data, dict):
+      if key is None:
+        raise ValueError("Images are stored as a dict, a key must be provided!")
+      try:
+        data = data[key]
+      except KeyError:
+        raise KeyError("Wrong key for provided data.")
+
+    if not isinstance(data, np.ndarray):
+      raise TypeError("Data must be stored as a dict or numpy array.")
+
+    self._initializeDimensions(data)
+
+    return data
+
+
+  def _initializeDimensions(self, inputData):
+    """
+    Stores the training images' dimensions, for convenience.
+    """
+    if len(inputData.shape) == 2:
+      self.imageHeight, self.numImages = inputData.shape
+      self.imageWidth, self.numChannels = None, None
+
+    elif len(inputData.shape) == 3:
+      self.imageHeight, \
+      self.imageWidth, \
+      self.numImages = inputData.shape
+      self.numChannels = None
+
+    elif len(inputData.shape) == 4:
+      self.imageHeight, \
+      self.imageWidth, \
+      self.numChannels, \
+      self.numImages = inputData.shape
+
+    else:
+      raise ValueError("The provided image set has more than 4 dimensions.")
+
+
+  def _getDataBatch(self, inputData):
+    """
+    Returns an array of dimensions (inputDim, batchSize), to be used as
+    batch for training data.
+
+    This implementation uses random sub-images from a random image as
+    training batch.
+
+    Images are flattened to get a 2-dimensional batch.
+    """
+    if not hasattr(self, 'numImages'):
+      self._initializeDimensions(inputData)
+
+    batch = np.zeros((self.inputDim, self.batchSize))
+    miniImageSize = int(np.sqrt(self.inputDim))
+    # choose random image
+    imageIdx = np.random.choice(range(self.numImages))
+
+    for i in xrange(self.batchSize):
+      # pick random starting row
+      rows = self.imageHeight - miniImageSize - 2 * self.BUFFER
+      rowNumber = self.BUFFER + np.random.choice(range(rows))
+
+      if self.imageHeight is None:
+        miniImage = inputData[rowNumber : rowNumber + miniImageSize,
+                              imageIdx]
+      else:
+        # pick random starting column
+        cols = self.imageWidth - miniImageSize - 2 * self.BUFFER
+        colNumber = self.BUFFER + np.random.choice(range(cols))
+
+        if self.numChannels is None:
+          miniImage = inputData[rowNumber : rowNumber + miniImageSize,
+                                colNumber : colNumber + miniImageSize,
+                                imageIdx]
+        else:
+          # pick random channel
```

Comment:
I think both would work (since we are sampling many of those mini images, sampling the channel each time). But I can also take all channels and flatten them. Then filters would also have 3 channels right?